### PR TITLE
Bucket Sort added in cpp

### PR DESCRIPTION
**Pull request by YASH01009**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

(Yes)

### Description

(WoC)

### Checklist

- [Y] I've read the contribution guidelines.
- [Y] I've checked the issue list before deciding what to submit.
- [N] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests

(Can u pls tell what exactly to do with README.md)

### CODE REVIEW

**Comment by charlie219** [comment]
@YASH01009 update the `README.md` under C-Plus-Plus folder under **Sorting**. You must add the link to your code.
For further details, refer to PR #1412

**Comment by YASH01009** [comment]
I have edited the file.

**Comment by YASH01009** [comment]
Should I again make a new pull request after committing changes to the same branch ?? 
Thanks !! I am a bit new to open source.

**Comment by charlie219** [comment]
@YASH01009, you don't need to make another PR for the same issue, the changes made will be reflected.

File: C-Plus-Plus/sort/bucket_sort.cpp [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+void bucket_sort(vector<int>& arr) {
+	int n = arr.size(), k = 10;
+	// create a bucket
+	vector<vector<int>> bucket(k);
+	for(int e : arr)
+	    bucket[e/k].push_back(e);
+	for(vector<int> vec : bucket)
+		sort(vec.begin(), vec.end());
+	int ptr = 0, i = 0;
+	// cpoy the elts back to arr
```
Comment (Kajol-Kumari) [inline C-Plus-Plus/sort/bucket_sort.cpp]: please correct the typo here

File: C-Plus-Plus/sort/bucket_sort.cpp [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+	vector<int> arr(n);
+	// taking ip
+	for(int i=0; i<n; i++)
+		cin >> arr[i];
+	bucket_sort(arr);
+	// taking ip
+	for(int i=0; i<n; i++)
+		cout << arr[i] << ' ';
+	cout << '\n';
+}
```
Comment (Kajol-Kumari) [inline C-Plus-Plus/sort/bucket_sort.cpp]: leave an extra blank space at the end of the file

**Review by Kajol-Kumari — CHANGES_REQUESTED** [review]
Hey @YASH01009 please revise your logic for bucket sort in this code as I am not getting the desire input.

For input 
```
4
9 2 1 10
```
I am getting `9 2 1 10` as output which shouldn't be the case. Also I can see that users can input restricted values only as the value of `k` is fixed here. Can we please make it generalised ?

File: C-Plus-Plus/sort/bucket_sort.cpp (hunk 1 above)
Comment (YASH01009) [reply C-Plus-Plus/sort/bucket_sort.cpp]: I corrected this typo. I also resolved the error for the aforementioned output.

File: C-Plus-Plus/sort/bucket_sort.cpp (hunk 2 above)
Comment (YASH01009) [reply C-Plus-Plus/sort/bucket_sort.cpp]: I left a blank line at the end. Coming to generalizing, as far as my knowledge is concerned, this is similar to counting sort in constraints, I mean based on the input values we decide the size of the bucket. If you want me still generalize, I will think of some way...

File: C-Plus-Plus/sort/bucket_sort.cpp [hunk 3]
```diff
@@ -0,0 +1,43 @@
+// Usecase : if all the elements are in a given range
```
Comment (HarshCasper) [inline C-Plus-Plus/sort/bucket_sort.cpp]: Use Multi-Line Comments instead of Single-Line Comments

File: C-Plus-Plus/sort/bucket_sort.cpp [hunk 4]
```diff
@@ -0,0 +1,43 @@
+// Usecase : if all the elements are in a given range 
+// as in the case we perform counting sort.
+// As in this example, assume all elements are between
+// 0 - 99. We select a fair bucket size, as in this case 
+// 10
+// average time complexity : O(nlog(n/k)), where k is #buckets
```
Comment (HarshCasper) [inline C-Plus-Plus/sort/bucket_sort.cpp]: Time and Space Complexity can be mentioned at the end of the Code

File: C-Plus-Plus/sort/bucket_sort.cpp [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+using namespace std;
+
+void bucket_sort(vector<int>& arr) {
+	int n = arr.size(), k = 10;
+	// create a bucket
+	vector<vector<int>> bucket(k);
+	for(int e : arr)
+	    bucket[e/k].push_back(e);
+	for(int i = 0; i < k; i++)
+		sort(bucket[i].begin(), bucket[i].end());
```
Comment (HarshCasper) [inline C-Plus-Plus/sort/bucket_sort.cpp:19]: Have a consistent indentation, throughout the Code

File: C-Plus-Plus/sort/bucket_sort.cpp [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+// 10
+// average time complexity : O(nlog(n/k)), where k is #buckets
+// worst case time complexity : O(nlogn)
+// space complexity : O(n)
+
+#include <bits/stdc++.h>
+using namespace std;
+
+void bucket_sort(vector<int>& arr) {
+	int n = arr.size(), k = 10;
```
Comment (HarshCasper) [inline C-Plus-Plus/sort/bucket_sort.cpp:13]: Use meaningful variable names. Code Quality is one of the important aspects here. Try using Snake Case or Camel Case to name your variables.

File: C-Plus-Plus/sort/bucket_sort.cpp [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+	vector<int> arr(n);
+	// taking ip
+	for(int i=0; i<n; i++)
+		cin >> arr[i];
+	bucket_sort(arr);
+	// taking ip
+	for(int i=0; i<n; i++)
+		cout << arr[i] << ' ';
+	cout << '\n';
+}
```
Comment (HarshCasper) [inline C-Plus-Plus/sort/bucket_sort.cpp:42]: Have a sample I/O to show the Users, how its being done.

**Review by HarshCasper — CHANGES_REQUESTED** [review]

File: C-Plus-Plus/sort/bucket_sort.cpp (hunk 6 above)
Comment (HarshCasper) [reply C-Plus-Plus/sort/bucket_sort.cpp:13]: I don't yet see usage of meaningful variable names.

File: C-Plus-Plus/sort/bucket_sort.cpp (hunk 5 above)
Comment (HarshCasper) [reply C-Plus-Plus/sort/bucket_sort.cpp:19]: No changes made for making the indentation more consistent.

**Review by HarshCasper — CHANGES_REQUESTED** [review]
Resolve the Merge Conflicts as well.

**Comment by HarshCasper** [comment]
I'm closing the PR due to no activity. If you would like to reopen this PR to be reviewed please join the [Slack](https://join.slack.com/t/tesseractcodi-bxq5968/shared_invite/zt-ju6c0nqq-KgkTvFyxjNVbDRBIjADanw) and drop me a message.

**Pull request closed without merging**